### PR TITLE
Stories/regex

### LINKED ISSUE #5: Allow regular expression matches instead of fixed URLs when adding mocks
i.e.

```
mockSingle(URL: "(.*)/shop/clothing", data: someData)
```

or even something like

```
mockSingle(URL: "(.*)/shop/clothing") { [matches:String] in ..... return someResponse; }
```

### PR DESCRIPTION
**Pull request by nap-sam-dean**
Fixes #5

I've created the idea of a RequestMatcher which SessionMocks will defer to in order to match requests - currently it's just simple string and regular expressions, but anything could implement the protocol to do more complicated request matching if needed.

### CODE REVIEW

File: Pod/Classes/NSURLSession/Matchers/SimpleRequestMatcher.swift [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+ This struct matches by matching the path to a regular expression and by 
+ comparing the methods for equality.
+*/
+struct SimpleRequestMatcher : RequestMatcher {
+    
+    let pathMatcher: NSRegularExpression
+    
+    let method: String
+    
+    init(url: NSURL, method: String) {
```
Comment (kerrmarin) [inline Pod/Classes/NSURLSession/Matchers/SimpleRequestMatcher.swift:21]: Should this also throw and just pass back the error from the other `init`?

File: Pod/Classes/NSURLSession/Matchers/SimpleRequestMatcher.swift [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+        try! self.init(expression: "^\(path)$", method: method)
+    }
+    
+    init(expression: String, method: String) throws {
+        try self.pathMatcher = NSRegularExpression(pattern: expression, options: NSRegularExpressionOptions.AnchorsMatchLines)
+        self.method = method
+    }
+    
+    func matches(request: NSURLRequest) -> Bool {
+        guard request.HTTPMethod == self.method else { return false }
```
Comment (kerrmarin) [inline Pod/Classes/NSURLSession/Matchers/SimpleRequestMatcher.swift:32]: :clap: :clap: :+1: :+1:

Should we enforce case insensitive compare?

File: Pod/Classes/NSURLSession/Matchers/SimpleRequestMatcher.swift [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+        
+        let path = request.URL?.absoluteString ?? ""
+        let options = NSMatchingOptions(rawValue: 0)
+        let range = NSMakeRange(0, (path as NSString).length)
```
Comment (kerrmarin) [inline Pod/Classes/NSURLSession/Matchers/SimpleRequestMatcher.swift]: You can use `path.characters.count` here

**Comment by kerrmarin** [comment]
Good stuff -- what's the objective going forward re. `NSURLConnection`? Are we aiming to maintain it post-1.0, or will we drop it before 1.0?

File: Pod/Classes/NSURLSession/Matchers/SimpleRequestMatcher.swift (hunk 1 above)
Comment (nap-sam-dean) [inline Pod/Classes/NSURLSession/Matchers/SimpleRequestMatcher.swift:21]: I wasn't sure - in the end I assumed that if a string has been escaped by NSRegularExpression itself, then it's probably fine?

File: Pod/Classes/NSURLSession/Matchers/SimpleRequestMatcher.swift (hunk 2 above)
Comment (nap-sam-dean) [inline Pod/Classes/NSURLSession/Matchers/SimpleRequestMatcher.swift:32]: Never even occurred to me - what you do think? I don't think I've ever seen HTTP methods in lowercase?

File: Pod/Classes/NSURLSession/Matchers/SimpleRequestMatcher.swift (hunk 3 above)
Comment (nap-sam-dean) [inline Pod/Classes/NSURLSession/Matchers/SimpleRequestMatcher.swift]: I wasn't sure - is that the same as `NSString.length` - I'm going to be putting it into an `NSRange` . . . . :(

File: Pod/Classes/NSURLSession/Matchers/SimpleRequestMatcher.swift (hunk 2 above)
Comment (nap-sam-dean) [inline Pod/Classes/NSURLSession/Matchers/SimpleRequestMatcher.swift:32]: Ah-ha. from RFC2616 Section 5.1.1 : `The Method token indicates the method to be performed on the resource identified by the Request-URI. The method is case-sensitive.`

File: Pod/Classes/NSURLSession/Matchers/SimpleRequestMatcher.swift (hunk 3 above)
Comment (kerrmarin) [inline Pod/Classes/NSURLSession/Matchers/SimpleRequestMatcher.swift]: We have [precedents for this in our projects](https://github.com/net-a-porter-mobile/XCTest-Gherkin/blob/master/Pod/Core/XCTestCase%2BGherkin.swift#L216) as well as [other open source projects](https://github.com/fastlane/snapshot/blob/14247491a792016f83ac7aa33effa77a3b416d31/lib/assets/SnapshotHelper.swift#L54). I think it should be fine.

File: Pod/Classes/NSURLSession/Matchers/SimpleRequestMatcher.swift (hunk 2 above)
Comment (kerrmarin) [inline Pod/Classes/NSURLSession/Matchers/SimpleRequestMatcher.swift:32]: Sounds good then.

**Comment by nap-sam-dean** [comment]
@kerrmarin FYI: I've created an issue to remove NSURLConnection from the pod.

**Comment by kerrmarin** [comment]
Looks good :+1:

**Pull request merged**